core/src/worker/workflow/mod.rs

```diff
@@ -1023,17 +1024,31 @@ struct BufferedTasks {
     /// supersede any old one.
     wft: Option<PermittedWFT>,
     /// For query only tasks, multiple may be received concurrently and it's OK to buffer more
-    /// than one - however they should be dropped if, by the time we try to process them, we
-    /// have already processed a newer real WFT than the one the query was targeting (otherwise
-    /// we'd return data from the "future").
+    /// than one - however they must all be handled before applying the next "real" wft (after the
+    /// current one has been processed).
     query_only_tasks: VecDeque<PermittedWFT>,
+    /// These are query-only tasks for the *buffered* wft, if any. They will all be discarded if
+    /// a buffered wft is replaced before being handled. They move to `query_only_tasks` once the
+    /// buffered task is taken.
+    query_only_tasks_for_buffered: VecDeque<PermittedWFT>,
 }
 
 impl BufferedTasks {
+    /// Buffers a new task. If it is a query-only task, multiple such tasks may be buffered which
+    /// all will be handled at the end of the current WFT. If a new WFT which would advance history
+    /// is provided, it will be buffered - but if another such task comes in while there is already
+    /// one buffered, the old one will be overriden, and all queries will be invalidated.
     fn buffer(&mut self, task: PermittedWFT) {
         if task.work.is_query_only() {
-            self.query_only_tasks.push_back(task);
+            if self.wft.is_none() {
+                self.query_only_tasks.push_back(task);
+            } else {
+                self.query_only_tasks_for_buffered.push_back(task);
+            }
         } else {
+            if self.wft.is_some() {
+                self.query_only_tasks_for_buffered.clear();
+            }
             let _ = self.wft.insert(task);
```

Comment:
Why is this OK -- couldn't `self.wft` have signals and updates that we must process?

Reply:
By definition it is stale and invalid now if there's a new incoming WFT